src/cargo/util/progress.rs

```diff
@@ -74,6 +74,113 @@ struct Format {
+enum TaskbarValue {
+    /// Do not output anything
+    None,
+    /// Remove progress
+    Remove,
+    /// Progress value 0-100
+    Value(f64),
+    /// Indeterminate state (no bar, just animation)
+    Indeterminate,
+    /// Progress value 0-100 in the error state
+    Error(f64),
+}
+
+enum ProgressOutput {
+    /// Print progress without a message
+    PrintNow,
+    /// Progress, message and taskbar progress
+    TextAndTaskbar(String, TaskbarValue),
+    /// Only taskbar progress, no message and no text progress
+    Taskbar(TaskbarValue),
+}
+
+impl TaskbarProgress {
+    #[cfg(test)]
+    fn new(enabled: bool) -> Self {
+        Self {
+            enabled,
+            error: false,
+        }
+    }
+
+    /// Creates a new `TaskbarProgress` from a cargo's config.
+    /// If not explicitly enabled or disabled, detect a supported terminal.
+    fn from_config(gctx: &GlobalContext) -> Self {
+        let enabled = match gctx.progress_config().taskbar {
+            Some(v) => v,
+            None => {
+                gctx.get_env("WT_SESSION").is_ok()
+                    || gctx.get_env("ConEmuANSI").ok() == Some("ON".into())
+            }
+        };
+
+        Self {
+            enabled,
+            error: false,
+        }
+    }
+
+    pub fn remove(&self) -> TaskbarValue {
+        if self.enabled {
+            TaskbarValue::Remove
+        } else {
+            TaskbarValue::None
+        }
+    }
+
+    pub fn value(&self, percent: f64) -> TaskbarValue {
+        match (self.enabled, self.error) {
+            (true, false) => TaskbarValue::Value(percent),
+            (true, true) => TaskbarValue::Error(percent),
+            (false, _) => TaskbarValue::None,
+        }
+    }
+
+    pub fn indeterminate(&self) -> TaskbarValue {
+        match (self.enabled, self.error) {
+            (true, false) => TaskbarValue::Indeterminate,
+            (true, true) => TaskbarValue::Error(100.0),
+            (false, _) => TaskbarValue::None,
+        }
+    }
+
+    pub fn error(&mut self) {
+        self.error = true;
+    }
+}
+
+impl std::fmt::Display for TaskbarValue {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // From https://conemu.github.io/en/AnsiEscapeCodes.html#ConEmu_specific_OSC
+        // ESC ] 9 ; 4 ; st ; pr ST
+        // When st is 0: remove progress.
+        // When st is 1: set progress value to pr (number, 0-100).
+        // When st is 2: set error state in taskbar, pr is optional.
+        // When st is 3: set indeterminate state, pr is ignored.
+        // When st is 4: set paused state, pr is optional.
+        let (state, progress) = match self {
+            Self::None => return Ok(()),
+            Self::Remove => (0, 0.0),
+            Self::Value(v) => (1, *v),
+            Self::Indeterminate => (3, 0.0),
+            Self::Error(v) => (2, *v),
+        };
+        write!(f, "\x1b]9;4;{state};{progress:.0}\x1b\\")
```

Comment:
According to [1](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html), [wiki](https://en.wikipedia.org/wiki/ANSI_escape_code) and [MS Learn](https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences) `<ST>` is `ESC \ (0x1B 0x5C)` or `BEL (0x7)`. Both are supported but `ESC \` is preferred

Reply:
Seems like it was a right choise, see [a recent patch comments](https://github.com/systemd/systemd/pull/34964/files#diff-59ed3cf1ea99efb36edcb76142b4da2ffe9588f4a9a3483408c03fc704672b00R30-R35) for systemd.